### ISSUE TITLE
Mechanism to implement correlation schemas

### ISSUE THREAD

**Issue by lenzip** [hep-statistics-serialization-standard/hep-statistics-serialization-standard#10]
I believe that the standard should envision a mechanism to specify (and modify) correlation schemas of uncertainties. While relying on common naming is fine, especially when combining different experiments, it requires a preemptive agreement. Also, it may change depending on physics based considerations.  For this reasons I believe a mechanism should be identified to explicitly specify correlation schemas, possibly post-facto.

**Comment by cburgard** [comment]
Thank you very much for bringing up this important point!
Indeed correlation schemes are extremely important.
Fortunately, from the large experiments (I'm mostly talking about ATLAS, CMS and LHCb, although this list is for sure non-exclusive), we have made the experience that the RooFit-approach of "correlating things with the same name" is generally very powerful and covers basically all of our use cases.

One prominent counter-example that would need to go beyond the name-idenification would be the partial correlations of parameters, but seeing how it is generally difficult to come up with the correlation numbers for such cases, and how such an approach would very much call into question the very idea of what a "parameter" is (if we have 2 parameters that are 99% correlated, do we have 1 or 2 parameters?), I would at least for the moment want to refrain from such attempts, especially since we haven't needed them in ~15 years of LHC combinations.

I think the most useful approach here would be something that is akin to namespaces, namely that each experiement (by convention) assigns namespaces to their parameters. 
In ATLAS Higgs combinations for example, this is routinely done by ensuring that all experimental nuisance parameters are named "ATLAS_*", but a more formal definition of what constitutes a namespace and perhaps even nested namespaces would for sure be useful.

Any suggestions, also ones counter to what I try to outline here, are of course highly welcome!

**Comment by alexander-held** [comment]
Partial correlations would mean multi-dimensional constraint terms in practice I assume. That should conceptually already work I think?

**Comment by cburgard** [comment]
> Partial correlations would mean multi-dimensional constraint terms in practice I assume. That should conceptually already work I think?

Yes, we already do that, e.g. when using regularized profile likelihood unfolding. If that's all that people are after, then yes, we have it and it works.

**Comment by jonas-eschle** [comment]
Good point, thanks for bringing this up! I think there are two things in the question, as already a bit picked out:
- how to have arbitrary correlations?
- how do we combine measurements?

Arbitrary correlations are an important tool, but they are technically covered if we assume that we can have arbitrary, multidimensional PDFs (and therefore likelihood terms, for the matter). So technically, there is a way.

Combining measurements can be a bit trickier: technically, I agree with @cburgard that the naming is the "only reasonable" way to go and appending something (this is technically trivial if you introduce namescopes in the loading/dumping/merger). What's trickier are shared parameters/constraints for example: if a parameter is i.e. taken from theory, this has to be the same parameter, even if two (or any number of) measurements are combined. Same goes for specific constraints: if measurement A constrains a parameter, measurement B may uses the same assumptions/theory input (although maybe formulated differently). These clearly need to be unique. But I think for this, there could be a way of standardizing some names, especially the ones that are used often and common, for the rest, that's probably up for each combination and hard/impossible to automatize.

**Comment by lmoureaux** [comment]
As far as I can tell, a tool for combining measurements would produce a new HS3 file to be consumed by other tools (e.g. for fitting). Such a tool would work as follows (say we have files `a.hf3` and `b.hf3`):
* Collect all parameters from both files
* Merge the ones that come with identical names and constraints (or implement smarter rules)
* Ask the user what to do about the rest, defaulting to fully independent parameters

So we'd have something like this that the user would customize until they're satisfied with the modeling (this example being awkward due to not well-thought naming conventions):
```
a.hdf3          b.hdf3          action (reason)
------------------------------------------------------------
a::lumi         b::lumi         merge (same name)
a::efficiency   b::efficiency   none (different constraints)
a::qcd_scale    (none)          none (no matching param)
(none)          b::mu_scale     none (no matching param)
```
As long as parameters can be merged and the combination tool can technically add any arbitrary constraints we can think of, is it really within the scope of HS3 to specify things such as naming conventions? I would say it belongs to another document because we want the format to survive as best practices evolve.

**Comment by jonas-eschle** [comment]
> As long as parameters can be merged and the combination tool can technically add any arbitrary constraints we can think of, is it really within the scope of HS3 to specify things such as naming conventions? I would say it belongs to another document because we want the format to survive as best practices evolve.

Maybe there was a misunderstanding, I think we completely agree: my point was exactly that we *should not* attempt to solve the problem of conventions: _Technically_, the problem is solved. Same name, same param. Conventions _can_ be adopted by whomever, someone who knows and cares more about it than us, (an inofficial recommendation from HS3 "why don't you prepend your libraries name?" is sufficient I think) but doesn't has to. It's fine for pyhf to add a "zfit_pdf", that's absolutely legal, as long as "zfit_pdf" is not a protected name.

And yes, I imagine a tool to do something like this, thumbs up for that!
